src/boot/efi/boot.c

```diff
@@ -1659,7 +1692,7 @@ static UINTN config_entry_find(Config *config, const CHAR16 *needle) {
         if (!needle)
                 return IDX_INVALID;
 
-        for (INTN i = config->entry_count - 1; i >= 0; i--)
+        for (UINTN i = 0; i < config->entry_count; i++)
```

Comment:
Now this is getting really silly: You sort versions by `sort-key` descending and the stuff without one ascending. But now you select the default entry by the resulting sort order. This is messed up.

This will select the topmost entry, which will be correct for `sort-key` but breaks compatibility for entries without one (it will pick older versions first).

Reply:
hmm, no? we sort:

1. if sort-key exists:
    1. by sort-key ascending
    2. by machine-id ascending
    3. by version descending
2. by id ascending

And that's exactly what we should do. And if we sorted things that way we should always search from beginning to end if we look for a default pattern or so.

Example:

1. sort-key=fedora version=35.1 machine-id=32dbce7a63f541c5b43d6265b713886f
2. sort-key=fedora version=35.2 machine-id=32dbce7a63f541c5b43d6265b713886f
3. sort-key=fedora version=36.1 machine-id=e3a1f5d8617c4ea7a66b560be6373b04
4. sort-key=fedora version=36.2 machine-id=e3a1f5d8617c4ea7a66b560be6373b04
5. sort-key=arch version=4711 machine-id=159064e98f7f4d38b577c2277c439bfa
6. sort-key=arch version=4712 machine-id=159064e98f7f4d38b577c2277c439bfa

This would result in the following order:

6, 5, 2, 1, 4, 3

And that looks absolutely appropriate and desired to me

Reply:
And then you have entries that lack a sort-key, and versions will be sorted 1 < 2 < 3 (instead of 3 < 2 < 1). This is inconsistent and unexpected and in-fact, breaks creates a regression of version 3 being preferred.